README.md
# NAME

Any::Moose - \*deprecated\* - use Moo instead!

# VERSION

version 0.24

# DEPRECATION

日本語翻訳: http://blog.64p.org/entry/2013/02/06/094906

Please use [Moo](https://metacpan.org/pod/Moo) instead of Any::Moose for new code.

Moo classes and roles will transparently and correctly upgrade to
Moose when needed. This is a fundamentally better design than what
Any::Moose offers. Mouse metaclasses do not interact with Moose
metaclasses which leaks abstractions all over the place.

Any::Moose had a good run. It was a simplistic but expedient answer
for getting Moose syntax on the cheap but with the transparent
upgrade path to Moose when you needed it. But really, please don't
use it any more. :)

You may find [MooX::late](https://metacpan.org/pod/MooX::late) useful for porting your code from
Any::Moose to Moo.

For the sparse documentation Any::Moose used to include, see
[https://metacpan.org/module/SARTAK/Any-Moose-0.18/lib/Any/Moose.pm](https://metacpan.org/module/SARTAK/Any-Moose-0.18/lib/Any/Moose.pm)

# AUTHORS

- Shawn M Moore <code@sartak.org>
- Florian Ragwitz <rafl@debian.org>
- Stevan Little <stevan@iinteractive.com>
- Tokuhiro Matsuno <tokuhirom@gmail.com>
- Goro Fuji <gfuji@cpan.org>
- Ricardo Signes <rjbs@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Best Practical Solutions.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

# CONTRIBUTORS

- Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
- Graham Knop <haarg@haarg.org>
- Karen Etheridge <ether@cpan.org>
- Stevan Little <stevan.little@iinteractive.com>
- gfx <gfuji@cpan.org>
- tokuhirom <tokuhirom@gmail.com>


Changes
Revision history for Any-Moose

{{$NEXT}}

0.27      2016-11-08 03:37:04Z
  - add deprecation warning when this module is used

0.26      2015-01-24 07:52:46Z
  - change Mouse $VERSION check to avoid dying on bad 'v2.4.0' version

0.25      2015-01-21 04:31:32Z
  - mark module as deprecated in metadata
  - add additional checks for too-old Mouse

0.24      2014-06-23 15:53:03Z
  - fixed bad changelog from last release, doh!

0.23      2014-06-23 13:04:35Z
  - now always prefers Moose if Moo is loaded, to overcome some
    incompatibility issues in subclasses (github #1, Graham Knop)

0.22      2014-06-19 02:30:13Z
  - fix missing =encoding directive in pod
  - canonical repository moved to https://github.com/moose/Any-Moose

0.21 2013-03-02
  - Add tokuhirom's translation of the deprecation notice

0.20 2012-12-30
  - Documentation fixes

0.19 2012-12-30
  - Deprecated in favor of Moo. Good night, sweet prince.

0.18 2011-11-07
  - Try to cope with broken smokers better by recording a Moose
    prereq when it is used. (Chris Williams)
    - see https://rt.cpan.org/Public/Bug/Display.html?id=72043

0.17 2011-09-16
  - Better indicate that this module is dynamically configured.
    - fixes [rt.cpan.org #70996]

0.16 2011-09-15
  - If you have Moose installed you no longer get the Mouse dependency (rjbs, reported by andk)
     Any::Moose has long had both runtime and test support for Moose
     without Mouse, but the toolchain has not, until recently, been
     good enough to handle alternatives. This change is designed
     to affect very few people.
    - fixes [rt.cpan.org #70597]

0.15 2011-06-07
  - Standardize the tests' Mouse/Moose loading eval/skip code
    (Sartak, reported by Daniel Macks)
    - fixes [rt.cpan.org #68657]

0.14 2011-04-29
  - Skip t/004-x-modules-mouse.t if you don't have MouseX::Types 0.03 since
    older versions do not work with modern Mouse (Sartak, reported by ewilhelm)

0.13  2010-05-19
  - Add load_first_existing_class (gfx)

0.12  2010-04-02
  - t/000-version.t for better diagnostics (tokuhirom)
  - Slight performance improvements for is_class_loaded,
      lazily loading Carp, etc (Sartak)
  - Start some real documentation (Sartak)
  - Document $ENV{ANY_MOOSE} (Sartak)
    - fixes [rt.cpan.org #52339]
  - Test that Moose is loaded, not CMOP (Sartak)
    - fixes [rt.cpan.org #56093]
  - Alias class_of and more functions (Sartak)
    - requested by [rt.cpan.org #52275]

0.11  2009-12-19
  - Support "use Any::Moose '::Util::TypeConstraints'" (gfx)
  - Support "no Any::Moose '::Util::TypeConstraints'" (gfx)
  - $@ related fixes (gfx)
  - Documentation and other minor improvements (gfx)
  - Any::Moose should now work when you have Moose but not Mouse. Still need to
    work out how to encode this in the installation system (Sartak)

0.10  2009-06-19
  - Fix META.yml

0.09  2009-05-19
  - Depend on Mouse 0.21.

0.08  2009-05-18
  - Add moose_is_preferred and mouse_is_preferred functions for when you
    need to care about which backend is in use (Sartak)
  - Deprecate the internal is_moose_loaded in favor of moose_is_preferred
    (Sartak)

0.07  2009-04-10
  - Depend on Mouse 0.20 for better compatibility with Moose >= 0.74 (rafl).

0.06  2009-03-18
  - Add Any::Moose::is_class_loaded. (tokuhirom)

0.05  2009-02-25
  - Stop using ambiguous ways of creating named subtypes in the tests. This
    makes them pass again with Moose >= 0.72. (rafl)

0.04  2009-02-17
  - Any::Moose now gives you strict and warnings (oops!)
  - Support "use Any::Moose 'X::Types'" (tokuhirom)

0.03  2009-02-09
  - When we load one of Mo*se, we'll continue to load that backer even if
    Moose had become loaded
  - Skip some tests if Moose is not installed
  - Hard dependency on Mouse

0.02  2009-02-08
  - Fix compilation error (rafl)
  - Add load_class and is_moose_loaded (stevan)

0.01  2009-02-05
  - Initial release



lib/Any/Moose.pm
package Any::Moose;
# ABSTRACT: (DEPRECATED) use Moo instead!
# KEYWORDS: deprecated Moose Mouse abstraction layer object-oriented
# vim: set ts=8 sts=4 sw=4 tw=115 et :

use 5.006_002;
use strict;
use warnings;

our $VERSION = '0.28';

warnings::warnif('deprecated', 'Any::Moose is deprecated. Please use Moo instead');

# decide which backend to use
our $PREFERRED;
do {
    local $@;
    if ($ENV{ANY_MOOSE}) {
        $PREFERRED = $ENV{'ANY_MOOSE'};
        warn "ANY_MOOSE is not set to Moose or Mouse"
            unless $PREFERRED eq 'Moose'
                || $PREFERRED eq 'Mouse';

        # if we die here, then perl gives "unknown error" which doesn't tell
        # you what the problem is at all. argh.
        if ($PREFERRED eq 'Moose' && !eval { require Moose }) {
            warn "\$ANY_MOOSE is set to Moose but we cannot load it";
        }
        elsif ($PREFERRED eq 'Mouse' && !eval { require Mouse }) {
            warn "\$ANY_MOOSE is set to Mouse but we cannot load it";
        }
    }
    elsif ( $INC{'Moo.pm'} && eval { require Moose } ) {
        $PREFERRED = 'Moose';
    }
    elsif (_is_moose_loaded()) {
        $PREFERRED = 'Moose';
    }
    elsif (eval { require Mouse; require Mouse::Util; Mouse::Util->can('get_metaclass_by_name') }) {
        $PREFERRED = 'Mouse';
    }
    elsif (eval { require Moose }) {
        $PREFERRED = 'Moose';
    }
    else {
        require Carp;
        warn "Unable to locate Mouse or Moose in INC";
    }
};

sub import {
    my $self = shift;
    my $pkg  = caller;

    # Any::Moose gives you strict and warnings
    strict->import;
    warnings->import;

    # first options are for Mo*se
    unshift @_, 'Moose' if !@_ || ref($_[0]);

    while (my $module = shift) {
        my $options = @_ && ref($_[0]) ? shift : [];

        $options = $self->_canonicalize_options(
            module  => $module,
            options => $options,
            package => $pkg,
        );

        $self->_install_module($options);
    }

    # give them any_moose too
    no strict 'refs';
    *{$pkg.'::any_moose'} = \&any_moose;
}

sub unimport {
    my $sel = shift;
    my $pkg = caller;
    my $module;

    if(@_){
        $module = any_moose(shift, $pkg);
    }
    else {
        $module = _backer_of($pkg);
    }
    my $e = do{
        local $@;
        eval "package $pkg;\n"
           . '$module->unimport();';
        $@;
   };

   if ($e) {
        require Carp;
        Carp::croak("Cannot unimport Any::Moose: $e");
   }

   return;
}

sub _backer_of {
    my $pkg = shift;

    if(exists $INC{'Mouse.pm'}
        and Mouse::Util->can('get_metaclass_by_name')
      ){
        my $meta = Mouse::Util::get_metaclass_by_name($pkg);
        if ($meta) {
            return 'Mouse::Role' if $meta->isa('Mouse::Meta::Role');
            return 'Mouse'       if $meta->isa('Mouse::Meta::Class');
       }
    }

    if (_is_moose_loaded()) {
        my $meta = Class::MOP::get_metaclass_by_name($pkg);
        if ($meta) {
            return 'Moose::Role' if $meta->isa('Moose::Meta::Role');
            return 'Moose'       if $meta->isa('Moose::Meta::Class');
        }
    }

    return undef;
}

sub _canonicalize_options {
    my $self = shift;
    my %args = @_;

    my %options;
    if (ref($args{options}) eq 'HASH') {
        %options = %{ $args{options} };
    }
    else {
        %options = (
            imports => $args{options},
        );
    }

    $options{package} = $args{package};
    $options{module}  = any_moose($args{module}, $args{package});

    return \%options;
}

sub _install_module {
    my $self    = shift;
    my $options = shift;

    my $module = $options->{module};
    (my $file = $module . '.pm') =~ s{::}{/}g;

    require $file;

    my $e = do {
        local $@;
        eval "package $options->{package};\n"
           . '$module->import(@{ $options->{imports} });';
        $@;
    };
    if ($e) {
        require Carp;
        Carp::croak("Cannot import Any::Moose: $e");
    }
    return;
}

sub any_moose {
    my $fragment = _canonicalize_fragment(shift);
    my $package  = shift || caller;

    # Mouse gets first dibs because it doesn't introspect existing classes

    my $backer = _backer_of($package) || '';

    if ($backer =~ /^Mouse/) {
        $fragment =~ s/^Moose/Mouse/;
        return $fragment;
    }

    return $fragment if $backer =~ /^Moose/;

    $fragment =~ s/^Moose/Mouse/ if mouse_is_preferred();
    return $fragment;
}

for my $name (qw/
    load_class
    is_class_loaded
    class_of
    get_metaclass_by_name
    get_all_metaclass_instances
    get_all_metaclass_names
    load_first_existing_class
        /) {
    no strict 'refs';
    *{__PACKAGE__."::$name"} = moose_is_preferred()
        ? *{"Class::MOP::$name"}
        : *{"Mouse::Util::$name"};
}

sub moose_is_preferred { $PREFERRED eq 'Moose' }
sub mouse_is_preferred { $PREFERRED eq 'Mouse' }

sub _is_moose_loaded { exists $INC{'Moose.pm'} }

sub is_moose_loaded {
    warnings::warnif('deprecated', "Any::Moose::is_moose_loaded is deprecated. Please use Any::Moose::moose_is_preferred instead");
    goto \&_is_moose_loaded;
}

sub _canonicalize_fragment {
    my $fragment = shift;

    return 'Moose' if !$fragment;

    # any_moose("X::Types") -> any_moose("MooseX::Types")
    $fragment =~ s/^X::/MooseX::/;

    # any_moose("::Util") -> any_moose("Moose::Util")
    $fragment =~ s/^::/Moose::/;

    # any_moose("Mouse::Util") -> any_moose("Moose::Util")
    $fragment =~ s/^Mouse(X?)\b/Moose$1/;

    # any_moose("Util") -> any_moose("Moose::Util")
    $fragment =~ s/^(?!Moose)/Moose::/;

    return $fragment;
}

1;
__END__

=head1 DEPRECATION

日本語翻訳: http://blog.64p.org/entry/2013/02/06/094906

Please use L<Moo> instead of Any::Moose for new code.

Moo classes and roles will transparently and correctly upgrade to
Moose when needed. This is a fundamentally better design than what
Any::Moose offers. Mouse metaclasses do not interact with Moose
metaclasses which leaks abstractions all over the place.

Any::Moose had a good run. It was a simplistic but expedient answer
for getting Moose syntax on the cheap but with the transparent
upgrade path to Moose when you needed it. But really, please don't
use it any more. :)

You may find L<MooX::late> useful for porting your code from
Any::Moose to Moo.

For the sparse documentation Any::Moose used to include, see
L<https://metacpan.org/module/SARTAK/Any-Moose-0.18/lib/Any/Moose.pm>

=cut
